File: java/isbn-verifier/src/main/java/IsbnVerifier.java
class IsbnVerifier {

	static final char VALID_CHECK_DIGIT = 'X';

	static boolean isValidIsbn(String stringToVerify) {
		if(!isValidCheckDigit(stringToVerify)) {
			return false;
		}

		String parsed = removeHyphens(stringToVerify);
		Integer value = computeValueForIsbn(parsed);
		return value % 11 == 0;
	}

	static Integer computeValueForIsbn(String isbn) {
		String reversed = new StringBuilder(isbn).reverse().toString();
		int value = 0;
		for (int i = 0; i < isbn.length(); i++) {
			char c = reversed.charAt(i);
			value += getValueFrom(c, i);
		}
		return value;
	}

	static Integer getValueFrom(char c, int i) {
		if(i == 0 && c == 'X') {
			return 10;
		}
		return Integer.parseInt(String.valueOf(c)) * (i + 1);
	}

	static String removeHyphens(String s) {
		return s.replaceAll("-", "");
	}

	static boolean isValidCheckDigit(String s) {
		char checkDigit = s.charAt((s.length() - 1));
		return Character.isDigit(checkDigit) || checkDigit == VALID_CHECK_DIGIT;
	}

    boolean isValid(String stringToVerify) {
		return isValidIsbn(stringToVerify);
    }

}

File: go/isbn-verifier/isbn_verifier.go
package isbn

import "unicode"

func IsValidISBN(isbn string) bool {
	if !isValidISBN(isbn) {
		return false
	}
	return false
}

func isValidISBN(isbn string) bool {
	if !isValidCheckDigit(rune(isbn[len(isbn)-1])) {
		return false
	}
	if !isValidPrefix(isbn) {
		return false
	}
	return true

}

func isValidCheckDigit(checkDigit rune) bool {
	return unicode.IsDigit(checkDigit) || checkDigit == 'X'
}

// Prefix in this context is the entire ISBN excluding the check digit
func isValidPrefix(isbn string) bool {
	for _, c := range isbn[0 : len(isbn)-1] {
		if c != '-' || unicode.IsDigit(c) {
			return false
		}
	}
	return true
}
